File: 백준/Gold/11000. 강의실 배정/강의실 배정.py
import sys
import heapq as hq

n = int(sys.stdin.readline())
meeting = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]

meeting.sort(key=lambda x: (x[0], x[1]))

rooms = []
hq.heappush(rooms, meeting[0][1])

for i in range(1, n):
    if meeting[i][0] >= rooms[0]:
        hq.heappop(rooms)
    hq.heappush(rooms, meeting[i][1])

print(len(rooms))

File: 백준/Gold/11000. 강의실 배정/강의실 배정.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        int n = Integer.parseInt(br.readLine());
        int[][] meeting = new int[n][2];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            meeting[i][0] = Integer.parseInt(st.nextToken());
            meeting[i][1] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(meeting, (a, b) -> {
            if (a[0] == b[0]) {
                return a[1] - b[1];
            }
            return a[0] - b[0];
        });

        PriorityQueue<Integer> rooms = new PriorityQueue<>();
        rooms.add(meeting[0][1]);

        for (int i = 1; i < n; i++) {
            if (meeting[i][0] >= rooms.peek()) {
                rooms.poll();
            }
            rooms.add(meeting[i][1]);
        }

        System.out.println(rooms.size());
    }
}
